src/derived.rs

```diff
@@ -370,16 +410,44 @@ where
 
         let mut map_write = self.map.write();
 
-        let changed_at = db.salsa_runtime().current_revision();
+        let current_revision = db.salsa_runtime().current_revision();
+        let changed_at = ChangedAt::Revision(current_revision);
 
         map_write.insert(
             key,
             QueryState::Memoized(Memo {
                 value: Some(value),
                 changed_at,
-                inputs: QueryDescriptorSet::new(),
-                verified_at: changed_at,
+                inputs: QueryDescriptorSet::default(),
+                verified_at: current_revision,
             }),
         );
     }
 }
+
+impl<DB, Q> Memo<DB, Q>
+where
+    Q: QueryFunction<DB>,
+    DB: Database,
+{
+    fn verify_memoized_value(&self, db: &DB) -> bool {
```

Comment:
For a tiny-bit less unwrappy API we can return an `Option<Q::Value>` instead of `bool` here.

Reply:
I considered that — I did not want to "consume" the memo entirely, but maybe that's silly. I could play around with it.

Reply:
We are cloning the value if it is valid, so we can clone it inside this method.

Reply:
OK, done